src/main/java/fi/aalto/cs/intellij/common/Course.java

```diff
@@ -0,0 +1,219 @@
+    try {
+      return new JSONObject(tokenizer);
+    } catch (JSONException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Course configuration file should consist of a valid JSON object", ex);
+    }
+  }
+
+  @NotNull
+  private static String getCourseName(@NotNull JSONObject jsonObject, @NotNull String path)
+      throws MalformedCourseConfigurationFileException {
+    try {
+      return jsonObject.getString("name");
+    } catch (JSONException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Missing or malformed \"name\" key", ex);
+    }
+  }
+
+  @NotNull
+  private static List<Module> getCourseModules(JSONObject jsonObject, String path)
+      throws MalformedCourseConfigurationFileException {
+    JSONArray modulesJsonArray;
+    try {
+      modulesJsonArray = jsonObject.getJSONArray("modules");
+    } catch (JSONException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Missing or malformed \"modules\" key", ex);
+    }
+
+    List<Module> modules = new ArrayList<>();
+    // Indexing loop used to simplify checking that each entry is a JSON object.
+    for (int i = 0; i < modulesJsonArray.length(); ++i) {
+      JSONObject moduleObject = modulesJsonArray.optJSONObject(i);
+      if (moduleObject == null) {
+        throw new MalformedCourseConfigurationFileException(path,
+            "\"modules\" value should be an array of objects", null);
+      }
+      modules.add(getIndividualModule(moduleObject, path));
+    }
+    return modules;
+  }
+
+  @NotNull
+  private static Module getIndividualModule(@NotNull JSONObject moduleObject, @NotNull String path)
+      throws MalformedCourseConfigurationFileException {
+    try {
+      String moduleName = moduleObject.getString("name");
+      URL moduleUrl = new URL(moduleObject.getString("url"));
+      return new Module(moduleName, moduleUrl);
+    } catch (JSONException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Module objects should contain \"name\" and \"url\" keys with string values", ex);
+    } catch (MalformedURLException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Malformed URL in module object", ex);
+    }
+  }
+
+  @NotNull
+  private static Map<String, String> getCourseRequiredPlugins(@NotNull JSONObject jsonObject,
+                                                              @NotNull String path)
+      throws MalformedCourseConfigurationFileException {
+    HashMap<String, String> requiredPlugins = new HashMap<>();
+    JSONObject requiredPluginsJson;
+    try {
+      requiredPluginsJson = jsonObject.getJSONObject("requiredPlugins");
+    } catch (JSONException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Missing or malformed \"requiredPlugins\" key", ex);
+    }
+    Iterable<String> keys = requiredPluginsJson::keys;
```

Comment:
a bit puzzled here, if the first `try-catch` throws `JSONException exception` and `requiredPluginsJson` is still uninitialized. what would then happen, if `requiredPluginsJson::keys;` is called? 
to my understanding, the IDE should be ruby-red about `JSONObject requiredPluginsJson` - a method variable is not initialized... or smth like that

Reply:
The `catch` block throws a new exception so the execution never goes that far.